internal/controlplane/handlers_reconciliationtasks.go

```diff
@@ -0,0 +1,102 @@
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package controlplane
+
+import (
+	"context"
+	"database/sql"
+	"errors"
+
+	"github.com/ThreeDotsLabs/watermill/message"
+	"github.com/google/uuid"
+	"github.com/rs/zerolog/log"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"github.com/stacklok/minder/internal/engine"
+	"github.com/stacklok/minder/internal/logger"
+	"github.com/stacklok/minder/internal/reconcilers"
+	pb "github.com/stacklok/minder/pkg/api/protobuf/go/minder/v1"
+)
+
+// CreateEntityReconciliationTask creates a task to reconcile the state of an entity
+func (s *Server) CreateEntityReconciliationTask(ctx context.Context,
+	in *pb.CreateEntityReconciliationTaskRequest) (
+	*pb.CreateEntityReconciliationTaskResponse, error,
+) {
+	// Populated by EntityContextProjectInterceptor using incoming request
+	entityCtx := engine.EntityFromContext(ctx)
+	err := entityCtx.Validate(ctx, s.store)
+	if err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "error in entity context: %v", err)
+	}
+
+	entity := in.GetEntity()
+
+	if entity == nil {
+		return nil, status.Error(codes.InvalidArgument, "entity is required")
+	}
+
+	var msg *message.Message
+	var topic string
+
+	// TODO: Support other entity types, replace with switch
+	if entity.GetType() == pb.Entity_ENTITY_REPOSITORIES {
+		msg, err = s.getRepositoryReconciliationMessage(ctx, entity)
+		if err != nil {
+			return nil, err
+		}
+		topic = reconcilers.InternalReconcilerEventTopic
+	} else {
+		return nil, status.Errorf(codes.InvalidArgument, "entity type %s is not supported", entity.GetType())
+	}
+
+	// This is a non-fatal error, so we'll just log it and continue with the next ones
+	if err := s.evt.Publish(topic, msg); err != nil {
+		log.Printf("error publishing reconciler event: %v", err)
+	}
+
+	return &pb.CreateEntityReconciliationTaskResponse{
+		TaskId: msg.UUID,
+	}, nil
+}
+
+func (s *Server) getRepositoryReconciliationMessage(ctx context.Context, entity *pb.EntityTypedId) (
+	*message.Message, error,
+) {
+	repoUUID, err := uuid.Parse(entity.GetId())
+	if err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "error parsing repository id: %v", err)
+	}
+
+	repo, err := s.store.GetRepositoryByID(ctx, repoUUID)
+	if errors.Is(err, sql.ErrNoRows) {
+		return nil, status.Errorf(codes.NotFound, "repository not found")
+	} else if err != nil {
+		return nil, status.Errorf(codes.Internal, "cannot read repository: %v", err)
+	}
+
```

Comment:
You could compare `repo.ProjectID` with `entityCtx.Project.ID` here to check the ownership (see my other comment).

Reply:
It should be possible to use the `GetRepositoryByIDAndProject` query on line 84

Reply:
Updated to use that